sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,625 @@
+       } else {
+         // TODO check for numberOfVersionsToRetain
+         // Destroy DB so that we can recontruct it using snapshot and delta files
+         RocksDbInstance.destroyDB(rocksDbPath)
+       }
+
+       // Check for snapshot files starting from "version"
+       while (!found && lastAvailableVersion > 0) {
+         found = {
+           try {
+             loadSnapshotFile(lastAvailableVersion)
+           } catch {
+             case e: Exception =>
+               logError(s"$e while reading snapshot file")
+               throw e
+           }
+         }
+         if (!found) {
+           lastAvailableVersion = lastAvailableVersion - 1
+         }
+         logInfo(s"Snapshot for $lastAvailableVersion for " +
+           s"partition ${stateStoreId_.partitionId} found = $found")
+       }
+
+       rocksDbWriteInstance = new OptimisticTransactionDbInstance(
+         keySchema, valueSchema, version.toString)
+        rocksDbWriteInstance.open(rocksDbPath, rocksDbConf)
+        rocksDbWriteInstance.startTransactions()
+
+       // Load all the deltas from the version after the last available
+       // one up to the target version.
+       // The last available version is the one with a full snapshot, so it doesn't need deltas.
+       for (deltaVersion <- (lastAvailableVersion + 1) to version) {
+         val fileToRead = deltaFile(baseDir, deltaVersion)
+         updateFromDeltaFile(fm, fileToRead, keySchema, valueSchema,
+           rocksDbWriteInstance, sparkConf)
+         logInfo(s"Read delta file for version $version of $this from $fileToRead")
+       }
+
+       rocksDbWriteInstance.commit(Some(getBackupPath(version)))
+       rocksDbWriteInstance.close()
+       rocksDbWriteInstance = null
+     }
+     catch {
+       case e: IllegalStateException =>
+         logError(s"Exception while loading state ${e.getMessage}")
+         if (rocksDbWriteInstance != null) {
+           rocksDbWriteInstance.abort()
+           rocksDbWriteInstance.close()
+         }
+         throw e
+     }
+   }
+    logInfo(s"Loading state for $version takes $elapsedMs ms.")
+  }
+
+  private def loadSnapshotFile(version: Long): Boolean = {
+    val fileToRead = snapshotFile(baseDir, version)
+    val tmpLocDir: File = new File(getTempPath(version))
+    val tmpLocFile: File = new File(getTempPath(version).concat(".tar"))
+    try {
+      if (!fm.exists(fileToRead)) {
+        return false
+      }
+      logInfo(s"Will download $fileToRead at location ${tmpLocFile.toString()}")
+      if (downloadFile(fm, fileToRead, new Path(tmpLocFile.getAbsolutePath), sparkConf)) {
+        FileUtility.unTar(tmpLocFile)
+        if (!tmpLocDir.list().exists(_.endsWith(".sst"))) {
+          logWarning("Snaphot files are corrupted")
+          throw new IOException(
+            s"Error reading snapshot file $fileToRead of $this:" +
+              s" No SST files found")
+        }
+        FileUtils.moveDirectory(tmpLocDir, new File(rocksDbPath))
```

Comment:
Can we make this configurable as in [here](https://github.com/apache/flink/blob/d8aac6ffb833d1d0348be0f6d280b465213d5df5/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java#L327)?

Reply:
Sure. Will make the changes.

Reply:
Done